config/confighttp/compression.go

```diff
@@ -68,25 +68,68 @@ func (r *compressRoundTripper) RoundTrip(req *http.Request) (*http.Response, err
 type decompressor struct {
 	errHandler func(w http.ResponseWriter, r *http.Request, errorMsg string, statusCode int)
 	base       http.Handler
+	decoders   map[string]func(body io.ReadCloser) (io.ReadCloser, error)
 }
 
 // httpContentDecompressor offloads the task of handling compressed HTTP requests
 // by identifying the compression format in the "Content-Encoding" header and re-writing
 // request body so that the handlers further in the chain can work on decompressed data.
 // It supports gzip and deflate/zlib compression.
-func httpContentDecompressor(h http.Handler, eh func(w http.ResponseWriter, r *http.Request, errorMsg string, statusCode int)) http.Handler {
+func httpContentDecompressor(h http.Handler, eh func(w http.ResponseWriter, r *http.Request, errorMsg string, statusCode int), decoders map[string]func(body io.ReadCloser) (io.ReadCloser, error)) http.Handler {
 	errHandler := defaultErrorHandler
 	if eh != nil {
 		errHandler = eh
 	}
-	return &decompressor{
+
+	d := &decompressor{
 		errHandler: errHandler,
 		base:       h,
+		decoders: map[string]func(body io.ReadCloser) (io.ReadCloser, error){
+			"": func(body io.ReadCloser) (io.ReadCloser, error) {
+				// Not a compressed payload. Nothing to do.
+				return nil, nil
+			},
+			"gzip": func(body io.ReadCloser) (io.ReadCloser, error) {
+				gr, err := gzip.NewReader(body)
+				if err != nil {
+					return nil, err
+				}
+				return gr, nil
+			},
+			"zstd": func(body io.ReadCloser) (io.ReadCloser, error) {
+				zr, err := zstd.NewReader(
+					body,
+					// Concurrency 1 disables async decoding. We don't need async decoding, it is pointless
+					// for our use-case (a server accepting decoding http requests).
+					// Disabling async improves performance (I benchmarked it previously when working
+					// on https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/23257).
+					zstd.WithDecoderConcurrency(1),
+				)
+				if err != nil {
+					return nil, err
+				}
+				return zr.IOReadCloser(), nil
+			},
+			"zlib": func(body io.ReadCloser) (io.ReadCloser, error) {
+				zr, err := zlib.NewReader(body)
+				if err != nil {
+					return nil, err
+				}
+				return zr, nil
+			},
+		},
+	}
+	d.decoders["deflate"] = d.decoders["zlib"]
```

Comment:
Would it be asking too much to add an explicit test for snappy, as this is the one failing for existing receivers?

Reply:
I could add one, but it would essentially be a duplicate of the "UnsupportedCompression" test here: https://github.com/open-telemetry/opentelemetry-collector/blob/ec033ca833d34108142c8a3bdf27ce2f2e517328/config/confighttp/compression_test.go#L185-L189

Reply:
My idea is that it would replace the "custom-encoding" test case with a real implementation, and it would be converted into a test that overrides an existing encoding once snappy is added. In any case, this isn't a blocker.

Reply:
I'll add this in a follow up PR since this is blocking contrib :)